apps/io-func-wallet-solution-backend/src/infra/attestation-service/android/assertion.ts

```diff
@@ -0,0 +1,216 @@
+import { createHash, createVerify } from "crypto";
+import { EmailString, NonEmptyString } from "@pagopa/ts-commons/lib/strings";
+import * as t from "io-ts";
+import { google, playintegrity_v1 } from "googleapis";
+import { JWK, KeyLike, exportSPKI, importJWK } from "jose";
+
+const ALLOWED_WINDOW_MILLIS = 1000 * 60; // 60 seconds
+
+export const GoogleAppCredentials = t.type({
+  type: NonEmptyString,
+  project_id: NonEmptyString,
+  private_key_id: NonEmptyString,
+  private_key: NonEmptyString,
+  client_email: EmailString,
+  client_id: NonEmptyString,
+  auth_uri: NonEmptyString,
+  token_uri: NonEmptyString,
+  auth_provider_x509_cert_url: NonEmptyString,
+  client_x509_cert_url: NonEmptyString,
+  universe_domain: NonEmptyString,
+});
+
+export type GoogleAppCredentials = t.TypeOf<typeof GoogleAppCredentials>;
+
+export type VerifyAssertionParams = {
+  integrityAssertion: NonEmptyString;
+  hardwareSignature: string;
+  clientData: string;
+  hardwareKey: JWK;
+  bundleIdentifier: string;
+  googleAppCredentials: GoogleAppCredentials;
+  androidPlayIntegrityUrl: string;
+  allowDevelopmentEnvironment: boolean;
+};
+
+export const playintegrity = google.playintegrity("v1");
+
+export const verifyAssertion = async (params: VerifyAssertionParams) => {
+  const {
+    integrityAssertion,
+    hardwareSignature,
+    clientData,
+    hardwareKey,
+    bundleIdentifier,
+    googleAppCredentials,
+    androidPlayIntegrityUrl,
+    allowDevelopmentEnvironment,
+  } = params;
+
+  // First check whether the clientData has been signed correctly with the hardware key
+  const signatureValidated = validateAssertionSignature(
+    hardwareKey,
+    clientData,
+    hardwareSignature
+  );
+
+  if (!signatureValidated) {
+    throw new Error("Invalid hardware signature");
+  }
+
+  // Then verify the integrity token
+  const jwtClient = new google.auth.JWT(
+    googleAppCredentials.client_email,
+    undefined,
+    googleAppCredentials.private_key,
+    [androidPlayIntegrityUrl]
+  );
+
+  google.options({ auth: jwtClient });
+  const result = await playintegrity.v1.decodeIntegrityToken({
+    packageName: bundleIdentifier,
+    requestBody: {
+      integrityToken: integrityAssertion,
+    },
+  });
+
+  const tokenPayloadExternal = result.data.tokenPayloadExternal;
+
+  if (!tokenPayloadExternal) {
+    throw new Error("Invalid token payload from Play Integrity API response");
+  }
+
+  const responseValidated = validateIntegrityResponse(
+    tokenPayloadExternal,
+    bundleIdentifier,
+    clientData,
+    allowDevelopmentEnvironment
+  );
+
+  if (!responseValidated) {
+    throw new Error("Integrity Response did not pass validation");
+  }
+
+  return true;
+};
+
+export const validateAssertionSignature = async (
+  hardwareKey: JWK,
+  clientData: string,
+  hardwareSignature: string
+) => {
+  const joseHardwareKey = (await importJWK(hardwareKey)) as KeyLike;
```

Comment:
Is there a way to avoid `as`?

Reply:
`importJWK`  return a `Promise<Uint8Array | KeyLike>` but exportSPKI require a `KeyLike`: `exportSPKI(key: KeyLike)`. 
The check could be done like this I think:

```js
const joseHardwareKey = await importJWK(hardwareKey);

if (typeof joseHardwareKey !== "object" || !("type" in joseHardwareKey)) {
  throw new Error("Invalid Hardware Key format");
}

const publicHardwareKeyPem = await exportSPKI(joseHardwareKey);
```

Reply:
yes sounds good

Reply:
Solved in [f7e7947](https://github.com/pagopa/io-wallet/pull/43/commits/f7e7947b620e2f76ab18240c6429eb25f94513f7)